File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/mlx/src/MLX.ts
/**
 * Apple MLX backend registration for the RunAnywhere React Native SDK.
 *
 * Public model lifecycle and inference APIs remain in `@runanywhere/core`.
 * This facade only connects the linked Swift MLX runtime to the commons
 * backend registry through core's existing Nitro bridge.
 */

import {
  isNativeModuleAvailable,
  requireNativeModule,
  SDKLogger,
} from '@runanywhere/core/internal';

const logger = new SDKLogger('MLX');
const DEFAULT_PRIORITY = 100;

export const MLX = {
  /** Version of the canonical Swift MLX runtime API. */
  version: '1.0.0',

  /** Register the linked Apple MLX runtime with the native backend registry. */
  async register(priority = DEFAULT_PRIORITY): Promise<boolean> {
    if (!Number.isFinite(priority)) {
      throw new RangeError('MLX backend priority must be a finite number');
    }
    if (!isNativeModuleAvailable()) {
      logger.warning(
        'Core native module not available; MLX registration skipped'
      );
      return false;
    }

    const native = requireNativeModule();
    if (!(await native.mlxRuntimeAvailable())) {
      logger.warning(
        'MLX runtime unavailable. Use a physical iOS device and ensure @runanywhere/mlx is installed with CocoaPods.'
      );
      return false;
    }

    const registered = await native.mlxRegisterBackend(priority);
    if (registered) {
      logger.info('MLX backend registered');
    } else {
      logger.warning('MLX backend registration returned false');
    }
    return registered;
  },

  /** Unregister the MLX runtime from the native backend registry. */
  async unregister(): Promise<boolean> {
    if (!isNativeModuleAvailable()) return false;
    const unregistered = await requireNativeModule().mlxUnregisterBackend();
    if (unregistered) {
      logger.info('MLX backend unregistered');
    }
    return unregistered;
  },

  /** Return whether the MLX runtime is currently registered. */
  async isRegistered(): Promise<boolean> {
    if (!isNativeModuleAvailable()) return false;
    return requireNativeModule().mlxIsBackendRegistered();
  },

  /**
   * Return whether a linked Apple MLX runtime can execute on this target.
   *
   * The iOS Simulator is intentionally unavailable; its arm64 slice supports
   * package, compile, and link validation only.
   */
  async isAvailable(): Promise<boolean> {
    if (!isNativeModuleAvailable()) return false;
    return requireNativeModule().mlxRuntimeAvailable();
  },
};

File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere_mlx/lib/mlx.dart
import 'dart:async';

import 'package:runanywhere/foundation/logging/sdk_logger.dart';
import 'package:runanywhere/native/types/basic_types.dart';
import 'package:runanywhere_mlx/src/native/mlx_bindings.dart';

/// Apple MLX backend module.
///
/// The implementation lives in the canonical Swift `RunAnywhereMLX` runtime.
/// This class installs that runtime's callback table into the C++ commons
/// registry. Models and inference are managed through the core `RunAnywhere`
/// APIs after registration succeeds.
abstract final class MLX {
  /// Version of the canonical Swift MLX runtime API.
  static const String version = '1.0.0';

  static bool _isRegistered = false;
  static MLXBindings? _bindings;
  static final SDKLogger _logger = SDKLogger('MLX');

  /// Registers the Apple MLX runtime with the shared backend registry.
  ///
  /// Returns `false` off iOS, in the iOS Simulator, or when the native runtime
  /// was not linked into the application. Calling this method more than once
  /// is safe.
  static Future<bool> register({int priority = 100}) async {
    if (isRegistered) {
      _logger.debug('MLX already registered');
      return true;
    }

    if (!isAvailable) {
      _logger.warning('MLX runtime is unavailable on this target');
      return false;
    }

    try {
      final bindings = _bindings ??= MLXBindings();
      final result = bindings.register(priority);
      if (result != RacResultCode.success &&
          result != RacResultCode.errorModuleAlreadyRegistered) {
        _logger.warning('MLX backend registration returned: $result');
        return false;
      }

      _isRegistered = true;
      _logger.info('MLX backend registered');
      return true;
    } catch (error) {
      _logger.warning('MLX registration failed: $error');
      return false;
    }
  }

  /// Unregisters the Apple MLX runtime.
  static Future<bool> unregister() async {
    if (!isRegistered) return true;

    final bindings = _bindings;
    if (bindings == null) return true;

    final result = bindings.unregister();
    if (result != RacResultCode.success) {
      _logger.warning('MLX backend unregistration returned: $result');
      return false;
    }

    _isRegistered = false;
    _logger.info('MLX backend unregistered');
    return true;
  }

  /// Whether the native runtime is registered with the shared registry.
  static bool get isRegistered {
    final bindings = _bindings;
    if (bindings == null) return _isRegistered;
    _isRegistered = bindings.isRegistered;
    return _isRegistered;
  }

  /// Whether a linked MLX runtime can execute on this target.
  ///
  /// MLX execution requires a physical iOS device. An arm64 simulator slice is
  /// packaged only so consumers can validate compilation and linkage.
  static bool get isAvailable => MLXBindings.checkAvailability();

  /// Starts registration without awaiting it.
  static void autoRegister() {
    unawaited(register());
  }

  /// Releases Dart-side binding state after unregistering the runtime.
  static Future<void> dispose() async {
    await unregister();
    _bindings = null;
    _isRegistered = false;
  }
}
